### PR TITLE
Migrate xen importer

### PR DESCRIPTION
**Pull request by TG1999**
Signed-off-by: Tushar Goel <tushar.goel.dav@gmail.com>

### CODE REVIEW

File: vulnerabilities/importers/xen.py [hunk 1]
```diff
@@ -13,60 +13,36 @@
...
-def fetch(url):
-    response = requests.get(url).content
-    return json.loads(response)
+    url = "https://xenbits.xen.org/xsa/xsa.json"
+    spdx_license_expression = ""
+    license_url = ""
+
+    def advisory_data(self):
+        data = fetch_response(self.url).json()
+        if not len(data):
```
Comment (pombredanne) [inline vulnerabilities/importers/xen.py:29]: What is the shape of data?
an example in the docstring would help a lot.
If this is a list, then `if not data` would be enough

File: vulnerabilities/importers/xen.py [hunk 2]
```diff
@@ -13,60 +13,36 @@
...
+            return []
+        xsas = data[0].get("xsas") or []
```
Comment (pombredanne) [inline vulnerabilities/importers/xen.py:31]: what is `data[0]`?  same as above ... an example would help.

File: vulnerabilities/tests/test_xen.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import json
```
Comment (pombredanne) [inline vulnerabilities/tests/test_xen.py:1]: This is missing our standard header

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thank you. The tests are failing and I have a few comments for your consideration.

File: vulnerabilities/importers/xen.py (hunk 2 above)
Comment (TG1999) [reply vulnerabilities/importers/xen.py:31]:
```suggestion
        xsas = data[0]["xsas"] or []
```

**Comment by TG1999** [comment]
License for this importer will be GPL-2 as stated here https://wiki.xenproject.org/wiki/Xen_FAQ_General

File: vulnerabilities/importers/xen.py [hunk 4]
```diff
@@ -7,66 +7,51 @@
...
+            yield from self.to_advisories(xsa)
+
+    def to_advisories(self, xsa):
+        xsa_id = xsa.get("xsa")
+        references = []
+        if xsa_id:
+            references.append(XsaReference.from_num_id(num_id=xsa_id))
+        title = xsa.get("title")
+        for cve in xsa.get("cve") or []:
+            if not is_cve(cve):
```
Comment (TG1999) [inline vulnerabilities/importers/xen.py:51]:
```suggestion
       #TODO: https://github.com/nexB/vulnerablecode/issues/981
            if not is_cve(cve):
```

**Review by TG1999 — COMMENTED** [review]
Suggestions:

File: vulnerabilities/references.py [hunk 5]
```diff
@@ -28,6 +28,16 @@ def from_id(cls, xsa_id):
             url=f"https://xenbits.xen.org/xsa/advisory-{numid}.html",
         )
 
+    @classmethod
+    def from_num_id(cls, num_id):
+        """
+        Return a new XsaReference from an XSA-XXXX id.
```
Comment (TG1999) [inline vulnerabilities/references.py:34]:
```suggestion
        Return a new XsaReference from an XSA number.
```

File: vulnerabilities/references.py [hunk 6]
```diff
@@ -28,6 +28,16 @@ def from_id(cls, xsa_id):
             url=f"https://xenbits.xen.org/xsa/advisory-{numid}.html",
         )
 
+    @classmethod
+    def from_num_id(cls, num_id):
```
Comment (TG1999) [inline vulnerabilities/references.py:32]:
```suggestion
    def from_number(cls, number):
```

File: vulnerabilities/references.py [hunk 7]
```diff
@@ -28,6 +28,16 @@ def from_id(cls, xsa_id):
...
+        """
+        Return a new XsaReference from an XSA-XXXX id.
+        """
+        return cls(
+            reference_id=f"XSA-{num_id}",
```
Comment (TG1999) [inline vulnerabilities/references.py:37]:
```suggestion
            reference_id=f"XSA-{number}",
```

File: vulnerabilities/importers/xen.py [hunk 8]
```diff
@@ -7,66 +7,51 @@
...
+            return []
+        xsas = data[0]["xsas"]
+        for xsa in xsas:
+            yield from self.to_advisories(xsa)
+
+    def to_advisories(self, xsa):
+        xsa_id = xsa.get("xsa")
+        references = []
+        if xsa_id:
+            references.append(XsaReference.from_num_id(num_id=xsa_id))
```
Comment (TG1999) [inline vulnerabilities/importers/xen.py:48]:
```suggestion
            references.append(XsaReference.from_number(number=xsa_id))
```

**Review by TG1999 — COMMENTED** [review]
Suggestions

**Review by pombredanne — APPROVED** [review]
LGTM

**Review by pombredanne — CHANGES_REQUESTED** [review]
We need to get a proper license

**Comment by johnmhoran** [comment]
LGTM 👍

**Review by johnmhoran — APPROVED** [review]

**Pull request merged**